src/number_theory/legendre_symbol/gauss_sum.lean

```diff
@@ -0,0 +1,315 @@
+        ← pow_apply' χ fp.1.pos (p ^ n), hχ.pow_char p], },
+end
+
+end gauss_sum_frob
+
+/-!
+### Values of quadratic characters
+-/
+
+section gauss_sum_values
+
+variables {R : Type u} [comm_ring R] [fintype R] {R' : Type v} [comm_ring R'] [is_domain R']
+
+/-- If the square of the Gauss sum of a quadratic character is `χ(-1) * #R`,
+then we get, for all `n : ℕ`, the relation `(χ(-1) * #R) ^ (p^n/2) = χ(p^n)`,
+where `p` is the (odd) characteristic of the target ring `R'`.
+This version can be used when `R` is not a field, e.g., `ℤ/8ℤ`. -/
+lemma char.card_pow_char_pow {χ : mul_char R R'} (hχ : is_quadratic χ) (ψ : add_char R R') (p n : ℕ)
+  [fp : fact p.prime] [hch : char_p R' p] (hp : is_unit (p : R)) (hp' : p ≠ 2)
+  (hg : (gauss_sum χ ψ) ^ 2 = χ (-1) * fintype.card R) :
+  (χ (-1) * fintype.card R) ^ (p ^ n / 2) = χ (p ^ n) :=
+begin
+  have : gauss_sum χ ψ ≠ 0,
+  { intro hf, rw [hf, zero_pow (by norm_num : 0 < 2), eq_comm, mul_eq_zero] at hg,
+    exact not_is_unit_prime_of_dvd_card p
+      ((char_p.cast_eq_zero_iff R' p _).mp $ hg.resolve_left (is_unit_one.neg.map χ).ne_zero) hp },
+  rw ← hg, apply mul_right_cancel₀ this,
+  rw [← quad_gauss_sum_frob_iter p n hp hχ ψ, ← pow_mul, mul_comm, ← pow_succ,
+      nat.two_mul_div_two_add_one_of_odd ((fp.1.eq_two_or_odd').resolve_left hp').pow],
+end
+
+/-- When `F` and `F'` are finite fields and `χ : F → F'` is a nontrivial quadratic character,
+then `(χ(-1) * #F)^(#F'/2) = χ(#F')`. -/
+lemma char.card_pow_card {F : Type} [field F] [fintype F] {F' : Type} [field F'] [fintype F']
+  {χ : mul_char F F'} (hχ₁ : is_nontrivial χ) (hχ₂ : is_quadratic χ)
+  (hch₁ : ring_char F' ≠ ring_char F) (hch₂ : ring_char F' ≠ 2) :
+  (χ (-1) * fintype.card F) ^ (fintype.card F' / 2) = χ (fintype.card F') :=
+begin
+  obtain ⟨n, hp, hc⟩ := finite_field.card F (ring_char F),
+  obtain ⟨n', hp', hc'⟩ := finite_field.card F' (ring_char F'),
+  let ψ := primitive_char_finite_field F F' hch₁,
+  let FF' := cyclotomic_field ψ.n F',
+  have hchar := algebra.ring_char_eq F' FF',
+  apply (algebra_map F' FF').injective,
+  rw [map_pow, map_mul, map_nat_cast, hc', hchar, nat.cast_pow],
+  simp only [← mul_char.ring_hom_comp_apply],
+  haveI := fact.mk hp',
+  haveI := fact.mk (hchar.subst hp'),
+  rw [ne, ← nat.prime_dvd_prime_iff_eq hp' hp, ← is_unit_iff_not_dvd_char, hchar] at hch₁,
+  exact char.card_pow_char_pow (hχ₂.comp _) ψ.char (ring_char FF') n' hch₁ (hchar ▸ hch₂)
+    (gauss_sum_sq (hχ₁.comp $ ring_hom.injective _) (hχ₂.comp _) ψ.prim),
+end
+
+end gauss_sum_values
+
+/-!
+### The quadratic character of 2
+
+This section proves the following result.
+
+For every finite field `F` of odd characteristic, we have `2^(#F/2) = χ₈(#F)` in `F`.
+This can be used to show that the quadratic character of `F` takes the value
+`χ₈(#F)` at `2`.
+
+The proof uses the Gauss sum of `χ₈` and a primitive additive character on `ℤ/8ℤ`;
+in this way, the result is reduced to `card_pow_char_pow`.
+-/
+
+section gauss_sum_two
```

Comment:
```suggestion
section gauss_sum_two

/-!
### The quadratic character of 2

This section proves the following result.

For every finite field `F` of odd characteristic, we have `2^(#F/2) = χ₈(#F)` in `F`.
This can be used to show that the quadratic character of `F` takes the value
`χ₈(#F)` at `2`.

The proof uses the Gauss sum of `χ₈` and a primitive additive character on `ℤ/8ℤ`;
in this way, the result is reduced to `card_pow_char_pow`.
-/
```

Reply:
FYI: all I'm doing is suggesting to put the module docstring inside the section.

Reply:
OK.